### ISSUE TITLE
Using clippy::missing_const_for_fn flags a function that has runtime behaviour

### ISSUE THREAD

**Issue by danwilliams** [rust-lang/rust-clippy#11874]
### Summary

Enabling the `clippy::missing_const_for_fn` lint (currently part of `clippy::nursery`) produces a lint warning for a constructor function that uses only runtime properties. The function accepts these properties in, and uses them to build an instance of `Self`. As far as I understand the nature of `const fn` functions, there is nothing here that is constant/static and therefore the result cannot be determined at compile time. I believe the triggering of this lint in this case is erroneous, i.e. a false positive.

### Lint Name

missing_const_for_fn

### Reproducer

I tried this code:

```rust
#[derive(Default)]
pub struct Settings {
	pub set_me: bool,
}

#[derive(Default)]
pub struct Foo {
	pub id:   Option<u64>,
	settings: Settings,
}

impl Foo {
	pub fn new(
		id:       Option<u64>,
		settings: Settings,
	) -> Self {
		Self {
			id,
			settings,
		}
	}
	
	pub const fn settings(&self) -> &Settings {
		&self.settings
	}
}
```

I saw this happen:

```
warning: this could be a `const fn`
   --> src/foo.rs:97:2
    |
97  | /     pub fn new(
98  | |         id:       Option<u64>,
99  | |         settings: Settings,
100 | |     ) -> Self {
...   |
104 | |         }
105 | |     }
    | |_____^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_const_for_fn
note: the lint level is defined here
   --> src/main.rs:57:2
    |
57  |     clippy::nursery,
    |     ^^^^^^^^^^^^^^^
    = note: `#[warn(clippy::missing_const_for_fn)]` implied by `#[warn(clippy::nursery)]`
```

I expected to see this happen:

No lint warning, i.e. the `new()` function should not trigger the lint. Note the `const fn settings()` function - if the `const` is removed here, this also triggers the lint, but this is correct in this case, as the function returns a borrow of the `settings` property, and that reference can be determined at compile time. However, the `Self` instance returned by `new()` is dynamic, and depends on the input parameters, and therefore I believe cannot be determined in the same way?

### Version

```text
rustc 1.74.0 (79e9716c9 2023-11-13)
binary: rustc
commit-hash: 79e9716c980570bfd1f666e3b16ac583f0168962
commit-date: 2023-11-13
host: x86_64-unknown-linux-gnu
release: 1.74.0
LLVM version: 17.0.4
```

**Comment by y21** [comment]
I don't think this is a false positive, or perhaps I'm misunderstanding this issue.

> However, the `Self` instance returned by `new()` is dynamic, and depends on the input parameters, and therefore I believe cannot be determined in the same way

The fact that this function depends on input parameters doesn't make it dynamic/runtime dependant.
If all input parameters are from a const context (which depends on where it's being called), then the function can also be evaluated in a const context.

As a user of this API, I can write:
```rs
const FOO: Foo = Foo::new(Some(42), Settings { set_me: true });
```
So marking `new` as a const fn is useful. If `new` wasn't a const fn, then I could not write this code.

**Comment by danwilliams** [comment]
@y21 thanks for the quick reply :)

I must admit I am sketchy about this issue, as I am not 100% au fait with how exactly the `const fn` scope is determined. However, before reporting the issue I considered the following points:

Function purpose: The `new()` function is a constructor for the `Foo` struct. It takes runtime parameters (`id` and `settings`) and returns a new instance of `Foo`. The purpose of this function is to initialize a struct with specific values at runtime. (Counter point: the compiler does know the nature of those things at compile time.)

Runtime data: My understanding is that `const fn` is limited to computations that can be done entirely at compile time. In this case, both `id` and `settings` are input parameters provided at runtime. Since `const fn` cannot depend on runtime inputs, this function cannot be `const`. (Counter point: Again, the compiler knows those things at compile time.)

Object instantiation: From my reading, `const fn` is not capable of creating new instances of objects with runtime data. The function's job is to instantiate `Foo` with potentially dynamic data, which is inherently a runtime operation.

Settings from outside: The fact that `settings` is an external parameter further cements the idea to me that this function operates with runtime data. Even if `id` could somehow be a constant, despite having multiple input possibilities, `settings` being a runtime parameter disqualifies the function from being `const`... doesn't it?

Multiple computable possibilities: This is the big one for me, as it's central to how I think I understand `const fn` functions to work. The example `fn a() { 23 }` is `const` because it always outputs `23`. The example `fn b(a: usize) { a }` can, I think, also be `const` (likely on a second-level pass after `a()` has been marked as `const`) because it can detect that `a()` is `const`. However, this situation always leads to the same output, because there is always the same input. Although a `const` function does not necessarily have to always produce the same output, I believe it has to be something that can be called and evaluated at compile-time, and has a deterministic output for compile-time inputs. In my example, both `id` and `settings` are passed in from multiple places, and therefore `new()` does not have a constant output, and neither can it be deterministically evaluated (well... see below). Does that not disqualify it from being `const`?

In the points above I am considering the "general" position. What I'm not sure about is how the specific situation of my contrived minimal example might affect each of those points.

In my example, I do not show the `new()` function being called. If it were called exactly once, then with the things passed in being an integer and a single struct instance containing a boolean, then on face value, that _might_ be an argument for saying that the input to the function is always static, and so the function can be `const`.

Except, the actual codebase this triggered in creates multiple instances of `Foo`, each using different input parameters. Although not shown in my example, I will happily expand the example to show this is it's a differentiating factor? Perhaps this does not matter, though, as the compiler knows each of these inputs in advance?

There is also one difference between the "real" code and the minimal example shown, that _might_ be important: the `settings` struct is more extensive. In the example I have condensed it to a single boolean, but in reality there is more stuff in there.

In short, I may have been overzealous in minimising the example code; perhaps I have obscured a critical element? But I am not sure exactly what the factors involved are.

Perhaps `const` functions are more clever than I am understanding/expecting... I am not using any conditionals, loops, or I/O in the execution chain, so theoretically I guess all of the pieces in this particular setup are known in advance. Does that mean my report is invalid?

**Comment by y21** [comment]
Perhaps the confusion comes from the fact that `const fn` and `const` items don't have the same semantics? A `const X = ...;` is always evaluated at compile time. `const fn`s however can execute at runtime.

>  Since `const fn` cannot depend on runtime inputs, this function cannot be `const`.

That's not entirely true.
`const fn` doesn't require that the function is used with compile time values. It only means that it *can* be used in a constant context with compile time values at callsite.
You can still call it with runtime values, in which case it's executed at runtime.

See for example, [`char::from_u32`](https://doc.rust-lang.org/stable/std/primitive.char.html#method.from_u32). It's a `const fn` and you can still call it with a value that you got at runtime.
Crucially, having it be a `const fn` means that it can *also* be used in constant context.

In other words, there's no harm in making something a `const fn` and it doesn't impose any restrictions for the caller.
It's strictly more useful because you can call it with both runtime values *and* in a const context, as opposed to a regular fn, which can only be called at runtime.


> The example `fn b(a: usize) { a }` can, I think, also be const (likely on a second-level pass after `a()` has been marked as const) because it can detect that `a()` is const

The relation between the caller of a function and the function itself doesn't matter for whether a function can be const or not (as in, whether `a` at callsite is a const or not).
What matters is that the function only does things that *can* be evaluated at compile time assuming that the parameter was a constant value.

**Comment by danwilliams** [comment]
@y21 Thank you for the explanation - that makes sense!

I think this is the key bit I was misunderstanding:

> A `const X = ...;` is always evaluated at compile time. `const fn`s however can execute at runtime.
> `const fn` doesn't require that the function is used with compile time values. It only means that it _can_ be used in a constant context with compile time values at callsite.
> You can still call it with runtime values, in which case it's executed at runtime.

I humbly apologise for my mistake here - thank you for taking the time to explain 👍 I had understood why `Foo.settings()` could become a `const fn`, but not realised that the key was not that it was dealing with a constant factor of the borrow, but rather, that it has behaviour that can be evaluated deterministically. That nuanced ability is what I had missed:

> ...you can call it with both runtime values _and_ in a const context, as opposed to a regular fn, which can only be called at runtime.

I think I need to carefully re-read the way that `const fn` works...

I'll close this ticket; clearly the lint works and is more clever than I am in detecting possible `const fn` cases! 😄

**Comment by y21** [comment]
No worries! Glad to be of help 😺
